src/xtopcom/xstate_sync/src/xstate_sync.cpp

```diff
@@ -437,19 +475,26 @@ evm_common::xh256_t xtop_state_sync::process_node_data(const xbytes_t & blob, st
     return res.Hash;
 }
 
-evm_common::xh256_t xtop_state_sync::process_unit_data(const xbytes_t & blob, std::error_code & ec) {
+evm_common::xh256_t xtop_state_sync::process_unit_data(const xbytes_t & blob, uint8_t version, std::error_code & ec) {
     evm_common::trie::SyncResult res;
-    base::xauto_ptr<base::xvbstate_t> bstate = base::xvblock_t::create_state_object({blob.begin(), blob.end()});
-    if (nullptr == bstate) {
-        ec = error::xerrc_t::state_data_invalid;
-        xerror("xtop_state_sync::process_unit_data hash: %s, data size: %zu, error %s", res.Hash.hex().c_str(), res.Data.size(), ec.message().c_str());
-        return {};
+    if (version == base::enum_xaccountindex_version_snapshot_hash) {
+        base::xauto_ptr<base::xvbstate_t> bstate = base::xvblock_t::create_state_object({blob.begin(), blob.end()});
+        if (nullptr == bstate) {
+            ec = error::xerrc_t::state_data_invalid;
+            xerror("xtop_state_sync::process_unit_data hash: %s, data size: %zu, error %s", res.Hash.hex().c_str(), res.Data.size(), ec.message().c_str());
+            return {};
+        }
+        auto unit_state = std::make_shared<data::xunit_bstate_t>(bstate.get());
+        auto snapshot = unit_state->take_snapshot();
+        auto state_hash = base::xcontext_t::instance().hash(snapshot, enum_xhash_type_sha2_256);
+        res.Hash = evm_common::xh256_t{xbytes_t{state_hash.begin(), state_hash.end()}};
+        res.Data = blob;
+    } else {
+        auto state_hash = base::xcontext_t::instance().hash({blob.begin(), blob.end()}, enum_xhash_type_sha2_256);
+        res.Hash = evm_common::xh256_t{xbytes_t{state_hash.begin(), state_hash.end()}};
```

Comment:
`res.Hash = evm_common::xh256_t{xspan_t<xbyte_t const>(state_hash.begin(), state_hash.end())};`
try it if it works.

Reply:
less copy